OpenBSD fix long socket addresses

semarie opened the pull request:
Original diff from @notgull in #118349, small changes from me.

on OpenBSD, getsockname(2) returns the actual size of the socket address, and  not the len of the content. Figure out the length for ourselves. see https://marc.info/?l=openbsd-bugs&m=170105481926736&w=2

Fixes #116523

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8643764942/job/23697463458) [(plain)](https://github.com/rust-lang/rust/commit/bd96a36a06ad069e69b00ed8c1a3842aa6cc7afb/checks/23697463458/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 28.9s done
#16 DONE 34.4s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
   Compiling addr2line v0.21.0
error[E0433]: failed to resolve: could not find `memchr` in `sys`
   --> library/std/src/os/unix/net/addr.rs:116:31
    |
116 |             len = crate::sys::memchr::memchr(0, sun_path)
    |                               ^^^^^^ could not find `memchr` in `sys`
help: consider importing this module
    |
1   + use core::slice::memchr;
    |
    |
help: if you import `memchr`, refer to it directly
    |
116 -             len = crate::sys::memchr::memchr(0, sun_path)
116 +             len = memchr::memchr(0, sun_path)

For more information about this error, try `rustc --explain E0433`.
error: could not compile `std` (lib) due to 1 previous error
Build completed unsuccessfully in 0:00:15
```

</details>

semarie committed:
OpenBSD fix long socket addresses

Original diff from @notgull in #118349, small changes from me.

on OpenBSD, getsockname(2) returns the actual size of the socket address, and 
not the len of the content. Figure out the length for ourselves.
see https://marc.info/?l=openbsd-bugs&m=170105481926736&w=2

Fixes #116523

Mark-Simulacrum commented:
@bors r+

bors commented:
:pushpin: Commit 7aaad6b7e2c04f97c1f9472dfeecea9662a35bb4 has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum 7aaad6b7e2c04f97c1f9472dfeecea9662a35bb4 -->
<!-- homu: {"type":"Approved","sha":"7aaad6b7e2c04f97c1f9472dfeecea9662a35bb4","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
